File: src/crypto.rs
#![allow(dead_code)]

use std::fmt::Write;

use anyhow::{anyhow, Result};
use blowfish::cipher::{block_padding::NoPadding, BlockDecryptMut, KeyIvInit};
use blowfish::Blowfish;
use cbc::Decryptor as BlowfishCbcDec;

#[cfg(test)]
use cbc::Encryptor as BlowfishCbcEnc;

const DEEZER_SECRET: &str = "g4el58wc0zvf9na1";
const CHUNK_SIZE: usize = 2048;
const BLOWFISH_BLOCK_SIZE: usize = 8;
const DEEZER_BLOWFISH_IV: [u8; 8] = [0, 1, 2, 3, 4, 5, 6, 7];

pub fn derive_blowfish_key(track_id: &str) -> [u8; 16] {
    let id_md5 = md5_hex(track_id.as_bytes());
    let id_md5_bytes = id_md5.as_bytes();
    let secret_bytes = DEEZER_SECRET.as_bytes();
    let mut key = [0u8; 16];

    for i in 0..16 {
        key[i] = id_md5_bytes[i] ^ id_md5_bytes[i + 16] ^ secret_bytes[i];
    }

    key
}

pub fn decrypt_audio_stream(track_id: &str, encrypted_bytes: &[u8]) -> Result<Vec<u8>> {
    let key = derive_blowfish_key(track_id);
    decrypt_audio_stream_with_key(&key, encrypted_bytes)
}

pub fn decrypt_chunked_stream(track_id: &str, encrypted_bytes: &[u8]) -> Result<Vec<u8>> {
    decrypt_audio_stream(track_id, encrypted_bytes)
}

pub fn decrypt_chunk_in_place_with_key(
    key: &[u8],
    chunk_index: usize,
    chunk: &mut [u8],
) -> Result<()> {
    if key.len() < 4 || key.len() > 56 {
        return Err(anyhow!("invalid Blowfish key length"));
    }

    if chunk_index % 3 != 0 {
        return Ok(());
    }

    let decryptable_len = chunk.len() - (chunk.len() % BLOWFISH_BLOCK_SIZE);
    if decryptable_len == 0 {
        return Ok(());
    }

    let chunk_prefix = &mut chunk[..decryptable_len];
    let cipher = BlowfishCbcDec::<Blowfish>::new_from_slices(key, &DEEZER_BLOWFISH_IV)
        .map_err(|_| anyhow!("failed to initialize Blowfish-CBC decryptor"))?;
    cipher
        .decrypt_padded_mut::<NoPadding>(chunk_prefix)
        .map_err(|_| anyhow!("failed to decrypt chunk using Blowfish-CBC"))?;

    Ok(())
}

pub fn decrypt_audio_stream_with_key(key: &[u8], encrypted_bytes: &[u8]) -> Result<Vec<u8>> {
    let mut output = Vec::with_capacity(encrypted_bytes.len());

    for (chunk_index, chunk) in encrypted_bytes.chunks(CHUNK_SIZE).enumerate() {
        let mut decrypted_chunk = chunk.to_vec();
        decrypt_chunk_in_place_with_key(key, chunk_index, &mut decrypted_chunk)?;
        output.extend_from_slice(&decrypted_chunk);
    }

    Ok(output)
}

fn md5_hex(bytes: &[u8]) -> String {
    let digest = md5::compute(bytes);
    let mut output = String::with_capacity(32);

    for byte in digest.0 {
        let _ = write!(&mut output, "{byte:02x}");
    }

    output
}

#[cfg(test)]
mod tests {
    use super::{
        decrypt_audio_stream_with_key, derive_blowfish_key, Blowfish, BlowfishCbcEnc, NoPadding,
        CHUNK_SIZE, DEEZER_BLOWFISH_IV,
    };
    use blowfish::cipher::{BlockEncryptMut, KeyIvInit};

    #[test]
    fn derived_key_is_stable() {
        let key = derive_blowfish_key("3135556");
        assert_eq!(hex::encode(key), "6c6c666b39662c37652575603c643439");
    }

    #[test]
    fn decrypts_only_every_third_chunk() {
        let key = derive_blowfish_key("42");
        let plaintext = vec![0x5a; CHUNK_SIZE * 4];
        let mut encrypted = plaintext.clone();

        for chunk_index in [0usize, 3usize] {
            let start = chunk_index * CHUNK_SIZE;
            let end = start + CHUNK_SIZE;
            let chunk = &mut encrypted[start..end];
            let cipher = BlowfishCbcEnc::<Blowfish>::new_from_slices(&key, &DEEZER_BLOWFISH_IV)
                .expect("valid key and IV");
            let msg_len = chunk.len();
            cipher
                .encrypt_padded_mut::<NoPadding>(chunk, msg_len)
                .expect("chunk is block aligned");
        }

        let decrypted = decrypt_audio_stream_with_key(&key, &encrypted).expect("decrypts");
        assert_eq!(decrypted, plaintext);
    }
}

File: go/internal/deezer/crypto.go
package deezer

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"encoding/hex"
	"errors"
	"fmt"

	"golang.org/x/crypto/blowfish"
)

const (
	deezerSecret      = "g4el58wc0zvf9na1"
	chunkSize         = 2048
	blowfishBlockSize = 8
	fieldSeparator    = "\u00a4"
)

var (
	deezerBlowfishIV = []byte{0, 1, 2, 3, 4, 5, 6, 7}
	cdnAESKey        = []byte("jo6aey6haid2Teih")
)

func DeriveBlowfishKey(trackID string) [16]byte {
	idMD5 := md5Hex([]byte(trackID))
	idMD5Bytes := []byte(idMD5)
	secretBytes := []byte(deezerSecret)
	var key [16]byte

	for i := 0; i < 16; i++ {
		key[i] = idMD5Bytes[i] ^ idMD5Bytes[i+16] ^ secretBytes[i]
	}

	return key
}

func DecryptAudioStream(trackID string, encrypted []byte) ([]byte, error) {
	key := DeriveBlowfishKey(trackID)
	return DecryptAudioStreamWithKey(key[:], encrypted)
}

func DecryptChunkedStream(trackID string, encrypted []byte) ([]byte, error) {
	return DecryptAudioStream(trackID, encrypted)
}

func DecryptChunkInPlaceWithKey(key []byte, chunkIndex int, chunk []byte) error {
	if len(key) < 4 || len(key) > 56 {
		return errors.New("invalid Blowfish key length")
	}
	if chunkIndex%3 != 0 {
		return nil
	}

	decryptableLen := len(chunk) - (len(chunk) % blowfishBlockSize)
	if decryptableLen == 0 {
		return nil
	}

	block, err := blowfish.NewCipher(key)
	if err != nil {
		return fmt.Errorf("initialize Blowfish-CBC decryptor: %w", err)
	}

	cipher.NewCBCDecrypter(block, deezerBlowfishIV).CryptBlocks(chunk[:decryptableLen], chunk[:decryptableLen])
	return nil
}

func DecryptAudioStreamWithKey(key []byte, encrypted []byte) ([]byte, error) {
	output := make([]byte, 0, len(encrypted))

	for chunkIndex, start := 0, 0; start < len(encrypted); chunkIndex, start = chunkIndex+1, start+chunkSize {
		end := start + chunkSize
		if end > len(encrypted) {
			end = len(encrypted)
		}

		decryptedChunk := append([]byte(nil), encrypted[start:end]...)
		if err := DecryptChunkInPlaceWithKey(key, chunkIndex, decryptedChunk); err != nil {
			return nil, err
		}
		output = append(output, decryptedChunk...)
	}

	return output, nil
}

func BuildCDNURL(md5Origin, songID string, format uint8, mediaVersion string) (string, error) {
	joined := fmt.Sprintf("%s%s%d%s%s%s%s", md5Origin, fieldSeparator, format, fieldSeparator, songID, fieldSeparator, mediaVersion)
	joinedHash := md5Hex([]byte(joined))
	payload := fmt.Sprintf("%s%s%s%s", joinedHash, fieldSeparator, joined, fieldSeparator)
	padded := []byte(payload)

	for len(padded)%aes.BlockSize != 0 {
		padded = append(padded, 0)
	}

	block, err := aes.NewCipher(cdnAESKey)
	if err != nil {
		return "", fmt.Errorf("initialize AES cipher: %w", err)
	}

	encrypted := make([]byte, len(padded))
	for i := 0; i < len(padded); i += aes.BlockSize {
		block.Encrypt(encrypted[i:i+aes.BlockSize], padded[i:i+aes.BlockSize])
	}

	if len(encrypted) == 0 {
		return "", errors.New("AES-ECB encryption produced an empty CDN payload")
	}

	return "https://f-cdnt-stream.dzcdn.net/mobile/1/" + hex.EncodeToString(encrypted), nil
}

func md5Hex(bytes []byte) string {
	sum := md5.Sum(bytes)
	return hex.EncodeToString(sum[:])
}
